### DISCUSSION TITLE
Why do PDMux Prefill and Decode kernels show no temporal overlap?

### DISCUSSION THREAD

**Discussion by ricklessboy404** [sgl-project/sglang#33652] [Q&A]
## Environment
* GPU: 3080Ti
* Model: Qwen3-1.7B
* CUDA Version: 12.8
* Pytorch Version: 2.9.1
* SGLang Version: 0.5.9

## Command
Launch a pdmux server:
```
nsys profile --trace-fork-before-exec=true --cuda-graph-trace=node -o pdmux.out --delay 30 \
python3 -m sglang.launch_server \
--model /root/autodl-tmp/models/qwen3_1.7B \
--tp 1 \
--disable-overlap-schedule \
--port 30000 \
--mem-fraction-static 0.75 \
--chunked-prefill-size -1 \
--enable-pdmux \
--pdmux-config-path pdmux_config.yml
```

Benchmark:
```
python3 -m sglang.bench_serving --backend sglang --host 127.0.0.1 --port 30000 --num-prompts 50 --request-rate 5
```

pdmux_config.yml:
```
# pdmux_config.yaml

# Number of SM groups to divide the GPU into.
# This includes two default groups:
#   - Group 0: all SMs for prefill
#   - Last group: all SMs for decode
# So the number of manual divisions (below) must be (sm_group_num - 2).
sm_group_num: 4

# Optional manual divisions of SMs.
# Each entry contains:
#   - prefill_sm: number of SMs allocated for prefill
#   - decode_sm: number of SMs allocated for decode
#   - decode_bs_threshold: minimum decode batch size at which this group should be selected
#
# If provided, the number of entries must equal (sm_group_num - 2).
# manual_divisions:
#   - [72, 8, 1]
#   - [104, 28, 5]
#   - [96, 36, 10]
#   - [80, 52, 15]
#   - [64, 68, 20]
#   - [56, 76, 25]


# Divisor for default stream index calculation.
# Used to adjust sm_group based on decode_bs when no manual_divisions are provided.
# Formula:
#   stream_idx = max(1, min(sm_group_num - 2, decode_bs * (sm_group_num - 2) // decode_bs_divisor))
decode_bs_divisor: 36

# Maximum token budget for split_forward in the prefill stage.
# Determines how many layers are executed per split_forward.
# Formula:
#   forward_count = max(1, forward_token_budget // extend_num_tokens)
split_forward_token_budget: 16384
```
## Phenomenon
I compared performance with and without `--enable-pdmux` and found that TTFT increases significantly when PDMux is turned on. To diagnose this, I profiled the PDMux-enabled configuration using `nsys`; the results are shown below.

<img width="2250" height="675" alt="result" src="https://github.com/user-attachments/assets/fb9530c2-27cc-4b8d-ad62-9e011784390a" />

<img width="1845" height="699" alt="Image" src="https://github.com/user-attachments/assets/eca6c129-af94-491d-8a3e-b176d7a6acb9" />

According to the nsys profile, Green Context 3 is assigned to Prefill and Green Context 4 to Decode. After the initial large-batch Prefill wave finishes, the two contexts appear to execute in a strictly alternating manner with almost no temporal overlap. 

## Question
Is this expected, or is there a known bottleneck (e.g., launch latency, stream synchronization, or SM scheduling) that prevents them from running concurrently?

**Comment by pm32900** [comment]
I think the behavior you're seeing can be expected, and `--disable-overlap-schedule` is probably not the direct reason for the lack of Prefill/Decode kernel overlap.

`--disable-overlap-schedule` controls SGLang's overlap between the CPU scheduler and the GPU model worker; it is different from the GPU-side concurrency that PDMux is trying to achieve with Green Contexts.

With PDMux, Prefill and Decode can be assigned different Green Contexts / SM partitions, but that does not guarantee that Nsight will show continuous simultaneous kernel execution. The contexts make concurrent execution possible by restricting the SM resources available to each workload, but actual overlap still depends on whether both sides have runnable work at the same time and on synchronization/dependencies between the split forwards.

In your configuration, I would specifically look at two things.

First, you have:

```text
split_forward_token_budget: 16384
```

PDMux splits the prefill forward pass according to this budget. If those split-forward regions are relatively coarse for Qwen3-1.7B on a 3080 Ti, Decode may only get scheduling opportunities around the boundaries between Prefill chunks. In Nsight this can look much more like alternating Prefill/Decode execution than continuous kernel-level overlap.

Second, the selected SM partition depends on `decode_bs`. With:

```text
sm_group_num: 4
decode_bs_divisor: 36
```

and no `manual_divisions`, the selected group follows the decode-batch-size-based mapping. Your workload (`request-rate 5`, 50 prompts) may therefore spend much of its time in a partition that is not particularly favorable for this GPU/model combination.

That can also explain why TTFT becomes worse: PDMux deliberately gives some GPU resources/scheduling opportunities to Decode while Prefill is active. On a relatively small model such as Qwen3-1.7B, the benefit from protecting Decode may not compensate for the reduced Prefill throughput and additional split-forward/scheduling overhead.

I would try three comparisons:

1. Profile without `--disable-overlap-schedule`. This should improve CPU/GPU scheduling overlap, although I would not expect it by itself to guarantee Prefill/Decode kernel overlap.

2. Reduce `split_forward_token_budget` and check whether Nsight shows finer-grained interleaving/overlap. If the current 16384 budget is producing long Prefill regions, a smaller value should provide more opportunities for Decode to run.

3. Try explicit `manual_divisions` for the 3080 Ti rather than relying on the default `decode_bs_divisor` mapping, and compare both TTFT and decode latency. The optimal split is workload- and GPU-dependent.

So I wouldn't interpret the trace as evidence that Green Contexts themselves are failing to provide isolation/concurrency. PDMux provides the resource partitioning needed for Prefill and Decode to coexist, but it doesn't imply that every Prefill kernel should visibly overlap a Decode kernel. The amount of actual overlap is determined by the workload, split-forward granularity, selected SM partition, and synchronization points.

The increased TTFT is therefore also plausible: you're trading some Prefill performance for better Decode isolation/concurrency, and that tradeoff can be unfavorable for this particular Qwen3-1.7B + 3080 Ti workload.

**Comment by flymysql** [comment]
## This is the intended design of PDMux, not a bug — and the alternation you're seeing is the *stream-group switching*, which necessarily synchronizes both streams

Your nsys reading is correct, and the reason is visible in the scheduler loop. Two things in the reply above are right but for the wrong reasons, and there's a serialization mechanism neither reply mentions. Also two traps specific to your setup (torch 2.9.1, and `--disable-overlap-schedule`) that are worth knowing about.

### 1. First, what PDMux actually is (it is not PD disaggregation)

The name is misleading, and it changes how to read the profile. PDMux is **intra-GPU SM partitioning via CUDA green contexts** — both prefill and decode run in *one process on one GPU*, on disjoint SM slices:

```python
# python/sglang/srt/arg_groups/fields/disagg.py
enable_pdmux: A[bool, "Enable PD-Multiplexing, PD running on greenctx stream."] = False
sm_group_num: A[int, "Number of sm partition groups."] = 8
```

and it's explicitly asserted NOT to be disaggregation:

```python
# python/sglang/srt/arg_groups/validation_hook.py
if cfg.enable_pdmux:
    assert cfg.pp_size == 1, "PD-Multiplexing is only supported with pipeline parallelism disabled (pp_size=1)."
    assert cfg.chunked_prefill_size == -1, "PD-Multiplexing is not compatible with chunked prefill."
    assert cfg.disaggregation_mode == "null", "PD-Multiplexing is not compatible with disaggregation mode."
    assert cfg.disable_overlap_schedule, "PD-Multiplexing is not compatible with overlap schedule."
```

So your `--chunked-prefill-size -1` and `--disable-overlap-schedule` are **required**, not suspicious. In particular, `--disable-overlap-schedule` disables *CPU-scheduler vs GPU-worker* overlap (a different axis from what you're profiling), which is why the first reply is right that it isn't the cause — but the real reason is subtler, below. And note PDMux is not PD: there's no KV transfer, no separate P/D instances; it's a single GPU split into SM slices that switch roles.

### 2. The green-context SM split only activates while *both* sides have work — and the transitions force a full sync

This is the core answer. Stream groups are built with a fixed layout: the first entry is full-GPU prefill, the last is full-GPU decode, and only the middle entries are real green-context partitions:

```python
# python/sglang/srt/multiplex/pdmux_context.py
SM_COUNTS.append((total_sm_count, 0))     # Normal stream for prefill  (all SMs)
SM_COUNTS.extend(divisions)               # green-context partitions
SM_COUNTS.append((0, total_sm_count))     # Normal stream for decode   (all SMs)
STREAM_GROUPS.append((torch.cuda.Stream(gpu_id), torch.cuda.Stream(gpu_id)))  # normal
for prefill_sm, decode_sm in divisions:
    STREAM_GROUPS.append(spatial.create_greenctx_stream_by_value(prefill_sm, decode_sm, gpu_id))
STREAM_GROUPS.append((torch.cuda.Stream(gpu_id), torch.cuda.Stream(gpu_id)))  # normal
```

Then the choice of group is made by:

```python
# python/sglang/srt/multiplex/multiplexing_mixin.py
def adjust_stream_groups(self, running_batch):
    if not running_batch.is_empty() and self.split_prefill_batch:
        # both have work -> pick a green-context partition (middle entries)
        set_current_stream_idx(stream_idx)
    elif not running_batch.is_empty():
        set_current_stream_idx(self.real_sm_group_num - 1)   # decode-only: ALL SMs
    else:
        set_current_stream_idx(0)                            # prefill-only: ALL SMs
    ...
    self.tp_worker.model_runner.update_decode_attn_backend(stream_idx)
```

**Green-context partitioning — the thing your nsys labels Green Context 3/4 — only happens when a running decode batch and a split prefill batch coexist.** When only one side has work, it grabs the whole GPU on normal streams. So the macro-pattern in your timeline is the engine cycling:

```
prefill-only (full GPU, normal streams) → both (green ctx split) → decode-only (full GPU, normal streams)
```

And **every one of those transitions synchronizes both streams**:

```python
if adjust_stream_group:
    prefill_stream.synchronize()
    decode_stream.synchronize()
    stream_idx, stream_group = self.adjust_stream_groups(running_batch=running_batch)
```

with `adjust_stream_group` being armed at each phase change:

```python
adjust_stream_group = created or adjust_stream_group        # a prefill batch was just created
...
adjust_stream_group = adjust_stream_group or (stream_idx > 0 and running_batch.is_empty())
...
self.split_prefill_batch = None
wait_prefill_kernel_done = False
adjust_stream_group = True                                  # the prefill just completed
```

So the "strictly alternating, almost no overlap" shape is **structural**: each alternation is a `synchronize()` on both streams, so any in-flight kernels on either stream must drain before the next phase begins. That directly produces the pattern you measured.

### 3. Two more serialization points inside every iteration

Even within a phase, the loop is not launch-and-continue:

- **Decode is synchronized every iteration**: `with torch.cuda.stream(decode_stream): decode_stream.synchronize(); if decode_done: self.process_batch_result(...)`. The CPU blocks on decode completion each time around.
- **Prefill completion is a CPU-side collective rendezvous**: when a split prefill finishes, it records an event, then does a blocking **host** allreduce to agree across TP ranks:

```python
prefill_exe_done_flag = prefill_exe_done.query()
flags = torch.ones(1, device="cpu", dtype=torch.int32) if prefill_exe_done_flag else torch.zeros(...)
self.tp_cpu_group.allreduce(flags, dist.ReduceOp.SUM).wait()
if flags.item() == get_parallel().tp_size:
    self.process_batch_result(self.split_prefill_batch, prefill_result)
    ...
```

plus `wait_prefill_kernel_done` deliberately holds prefill back for an iteration. The stated purpose of all of this is to **hide prefill latency in decode's shadow** and to keep prefill from delaying decode — i.e. it optimizes *interleaved* latency, not simultaneous throughput. The code even labels itself:

```python
# TODO(jason-fxz): this is a temporary demo
```

### 4. Two traps specific to your setup

**(a) Your torch version is in the documented degradation range.** The validation code warns:

> *"NOTE: CUDA Green Context may encounter potential issues with CudaGraph on torch 2.7.x – 2.8.x, leading to performance degradation."*
> *"...Please manually install torch 2.6.x."*

You're on **torch 2.9.1** — beyond the stated range, and the comment's concern (green context × CUDA graph) is exactly the interaction your `--cuda-graph-trace=node` profile is capturing. I'd treat graph+greenctx interaction as a live suspect for the *quantitative* part of your result (no overlap at all, rather than partial overlap). Trying torch 2.6.x, or disabling CUDA graphs, is a cheap way to separate "structural serialization" from "graph replay blocking on the green-context streams."

**(b) If you switch to `manual_divisions`, note the selection logic has no reset branch:**

```python
for i in range(len(manual_divisions)):
    _, _, threshold = manual_divisions[i]
    if decode_bs >= threshold:
        stream_idx = i + 1
```

There's no `else`: if `decode_bs` falls below *every* threshold, `stream_idx` keeps its previous value rather than falling back to the smallest partition. With your thresholds starting at high decode batch sizes, a small decode batch won't re-select. (You have `manual_divisions` commented out, so you're on the `decode_bs_divisor` auto-path — but this is worth knowing before you tune the manual table.)

### 5. So: expected, and what to do about it

**Yes, this is expected behavior for PDMux**, not a fabric/SM-scheduling bug and not a stream-synchronization defect. The policy is: one side gets the whole GPU whenever it's alone, and when both are active they get disjoint SM slices — with a mandatory both-streams barrier at every handoff. TTFT going up with PDMux enabled is the intended cost: prefill is being deliberately throttled/interleaved so decode isn't starved.

The knobs you have control the **granularity** of interleaving, not the amount of overlap:

- `split_forward_token_budget` (yours: 16384) — controls `forward_count = max(1, budget // extend_num_tokens)`, i.e. how many layers of a split prefill are launched per iteration. Larger → longer prefill chunks → fewer handoffs (less sync overhead) but coarser decode interference. This is the first thing I'd sweep.
- `decode_bs_divisor` (yours: 36) — controls which SM partition is selected as `decode_bs` grows.
- `sm_group_num` (yours: 4) — number of partition steps.

And a concrete expectation to set: **you will not tune this into full overlap**, because the design synchronizes at every phase boundary. If you need simultaneous prefill+decode throughput on one GPU, that isn't what PDMux provides — genuine concurrency is what disaggregation (`--disaggregation-mode` on separate instances) is for, which PDMux explicitly asserts against.

For a decisive measurement: profile with `--cuda-graph-trace=node` removed (or CUDA graphs disabled) and compare. If overlap appears without graphs, your "no overlap" is the greenctx×graph degradation the code warns about; if it stays strictly alternating, it's the phase-boundary sync in §2, which is by design.

*Verified against `main` on 2026-09-25: `arg_groups/fields/disagg.py` and `validation_hook.py` (pdmux constraints, torch-version warning), `multiplex/multiplexing_mixin.py` (`event_loop_pdmux`, `adjust_stream_groups`, per-iteration `decode_stream.synchronize()`, TP CPU allreduce), `multiplex/pdmux_context.py` (`initialize_stream_groups`, `SM_COUNTS` layout, `get_arch_constraints`), `distributed/parallel_state.py` (`pdmux_prefill_tp_group`).*
